RFCs/0004-pluggable-monitor.md

```diff
@@ -0,0 +1,453 @@
+
+1. the Client/IDE must first listen to a random TCP port (let's suppose it choose `32123`)
+1. the Client/IDE sends the command `OPEN 127.0.0.1:32123 /dev/ttyACM0` to the monitor tool
+1. the monitor tool opens `/dev/ttyACM0`
+1. the monitor tool connects via TCP/IP to `127.0.0.1:32123` and start streaming data back and forth
+
+The answer to the `OPEN` command is:
+
+```JSON
+{
+  "event": "open",
+  "message": "ok"
+}
+```
+
+If the monitor tool cannot communicate with the board, or if the tool can not connect back to the TCP port, or if any other error condition happens:
+
+```JSON
+{
+  "event": "open",
+  "error": true,
+  "message": "unknown port /dev/ttyACM23"
+}
+```
+
+The board port will be opened using the parameters previously set through the `CONFIGURE` command.
+
+Once the port is opened, it may be unexpectedly closed at any time due to hardware failure, or because the Client/IDE closes the TCP/IP connection. In this case an asynchronous `port_closed` message must be generated from the monitor tool:
+
+```JSON
+{
+  "event": "port_closed",
+  "message": "serial port disappeared!"
+}
+```
+
+or
+
+```JSON
+{
+  "event": "port_closed",
+  "message": "lost TCP/IP connection with the client!"
+}
+```
+
+#### CLOSE command
+
+The `CLOSE` command will close the currently opened port and close the TCP/IP connection used to communicate with the Client/IDE. The answer to the command is:
+
+```JSON
+{
+  "event": "close",
+  "message": "ok"
+}
+```
+
+or in case of error
+
+```JSON
+{
+  "event": "close",
+  "error": true,
+  "message": "port already closed"
+}
+```
+
+#### QUIT command
+
+The `QUIT` command terminates the monitor. The response to `QUIT` is:
+
+```JSON
+{
+  "eventType": "quit",
+  "message": "OK"
+}
+```
+
+after this output the monitor exits. This command is supposed to always succeed.
+
+#### Invalid commands
+
+If the client sends an invalid or malformed command, the monitor should answer with:
+
+```JSON
+{
+  "eventType": "command_error",
+  "error": true,
+  "message": "Unknown command XXXX"
+}
+```
+
+#### Reference implementations
+
+TODO...
+
+### Integration with `arduino-cli` and core platforms
+
+In this section we will see how monitors are distributed and integrated with Arduino platforms.
+
+#### Monitor tool distribution
+
+The monitor tools must be built natively for each OS and the CLI should run the correct tool for the running OS.
+
+The distribution infrastracture is already available for platform tools, like compilers and uploaders, through `package_index.json` so, the most natural way forward is to distribute also the monitor tools in the same way.
+3rd party developers should provide their monitor tools by adding them as resources in the `tools` section of `package_index.json` (at the `packages` level).
+
+Let's see how this looks into a `package_index.json` example:
```

Comment:
```suggestion
The distribution infrastructure is already available for platform tools, like compilers and uploaders, through [the Arduino package index](https://arduino.github.io/arduino-cli/latest/package_index_json-specification) so, the most natural way forward is to distribute also the monitor tools in the same way.
3rd party developers should provide their monitor tools by adding them as resources in the `tools` section of their package index (at the `packages` level).

Let's see an example of adding a monitor tool to a package index:
```
I think it's more clear to refer to it as "Arduino package index" rather than "`package_index.json`" because only the primary official package index has this filename but this system applies to all package indexes. This is the approach taken in the Arduino CLI documentation, including the package index specification.